Add allowControlWithoutGm setting

Larkinabout opened the pull request:
- Add allowControlWithoutGm setting to allow chosen user roles to control ATL effects while the GM is not present.
- Add info notification to advise user when ATL effects are disabled.

kaelad02 commented:
Any idea why the GM check is there in the first place? I could understand if there were socket calls to the GM then locking those features behind a GM check would be necessary, but I don't see that. I don't see any socket calls to the GM to do work on behalf of regular players. To be honest, I'd rather remove the GM check than add this setting.

Larkinabout commented:
I figured it was a security measure to stop players messing with lights behind the GM's back, which would make sense in some situations. I could expect some GMs wanting to keep that feature.

kaelad02 commented on src/activeLighting.js:
```diff
@@ -71,6 +71,21 @@ class ATL {
         });
+        game.settings.register("ATL", "allowControlWithoutGm", {
+            name: "Allow Control without GM",
+            hint: "Allow users with this role and above to control ATL effects while the GM is not present (Requires reload)",
```
Can you remove the reload suggestion? I think we should leave it up to Foundry or modules to indicate these now that Foundry can do the reload for us.
```suggestion
            hint: "Allow users with this role and above to control ATL effects while the GM is not present",
```

Larkinabout replied:
Fixed in latest commit.

Larkinabout replied:
I should note, I left the text on the hint to indicate to the user that it'll require a reload to take effect as they can still click to not reload, which won't do anything.

kaelad02 commented on src/activeLighting.js:
```diff
@@ -71,6 +71,21 @@ class ATL {
+                1: 'Player'
+            },
+            default: 4,
+            onChange: foundry.utils.debounce(() => window.location.reload(), 100)
```
Let's use Foundry's new v10 feature to do the reload for us.
```suggestion
            requiresReload: true
```

Larkinabout replied:
Fixed in latest commit.

kaelad02 commented on src/activeLighting.js:
```diff
@@ -71,6 +71,21 @@ class ATL {
+            hint: "Allow users with this role and above to control ATL effects while the GM is not present (Requires reload)",
+            scope: "world",
+            config: true,
+            type: String,
```
Why `String` when we're storing a number? Did using `Number` not work?

Larkinabout replied:
Fixed in latest commit.

kaelad02 commented on src/activeLighting.js:
```diff
@@ -71,6 +71,21 @@ class ATL {
+            config: true,
+            type: String,
+            choices: {
+                4: 'Game Master',
```
Can you change these to use double quotes to be consistent? Thanks

Larkinabout replied:
Fixed in latest commit.

Larkinabout commented:
I added in a hook for userConnected in the latest commit as attempting to only trigger the check on 'ready' meant players would still have permission to turn lights on after the GM left. I think the main trigger on 'ready' is still the right choice given it avoids rechecking permission for every update of an effect.

I also added in a bypass to allow players to turn off their lights without the GM present as I couldn't see any reason why that would be unwanted and it avoids double info messages when you try to turn a light on as a player and then turn it off again.

kaelad02 commented on src/activeLighting.js:
```diff
@@ -84,43 +99,71 @@ class ATL {
+    static async ready() {
+        async function getAllowed () {
+            if (game.user.isGM) return true;
+            const gm = !!game.users.find((u) => u.isGM && u.active);
```
One last thing...
```suggestion
            const gm = game.users.some((u) => u.isGM && u.active);
```

kaelad02 reviewed (commented):
Adding the `userConnected` hook seems like a reasonable addition. Saw one small thing. Fix that and I'll merge it in. There was a small, older issue that I'm gonna fix too and then I'll cut a release and ask Kandashi to post the listing to Foundry.
